libfwupd/fwupd-client.c

```diff
@@ -4240,6 +4250,54 @@ fwupd_client_upload_bytes_finish (FwupdClient *self, GAsyncResult *res, GError *
 	return g_task_propagate_pointer (G_TASK(res), error);
 }
 
+/* this is bad; we dlopen libsoup-2.4.so.1 and get the gtype manually
+ * to avoid deps on both libcurl and libsoup whilst preserving ABI */
+static void
+fwupd_client_ensure_soup_session (FwupdClient *self)
+{
+	FwupdClientObjectNewFunc func = NULL;
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	GType soup_gtype;
+
+	/* already set up */
+	if (priv->soup_session != NULL)
+		return;
+
+	/* known GType, just create */
+	soup_gtype = g_type_from_name ("SoupSession");
+	if (soup_gtype != 0) {
+		priv->soup_session = g_object_new (soup_gtype, NULL);
+		return;
+	}
+
+	/* load the library at rumtime, leaking the module */
+	if (priv->soup_module == NULL) {
+		g_autofree gchar *fn = NULL;
+		const gchar *libdirs[] = {
+			FWUPD_LIBDIR,
+			"/usr/lib64",
+			"/usr/lib",
+			NULL };
```

Comment:
hardcoded paths?  Is there a cleaner way to do this?

Reply:
I originally just had FWUPD_LIBDIR but that's prefixed with the, well, prefix and then it breaks for when you're developing fwupd. Perhaps that's actually okay, and we just use FWUPD_LIBDIR which will be the right thing when installed systemwide.

Reply:
What I was thinking was a `fu_common_get_path` type of solution, which will let this work well in the snap too, and also for developing fwupd.

Reply:
fu_common_get_path is libfwupdplugin, not libfwupd  :(

Reply:
ah..  Well in that case I guess yeah just make `FWUPD_LIBDIR` only.  Need to keep reminding myself it's just for gnome-software support, not general.  Snap-store will just need move at the same time they upgrade their library.

And `libfwupd3` can't come soon enough :)

Reply:
I actually think this would be fine for snap:

		tmp = g_getenv ("SNAP");
		if (tmp != NULL) {
			g_ptr_array_add (libdirs,
					 g_build_filename (tmp, FWUPD_LIBDIR, NULL));
		}

Reply:
although, maybe that's not required at all if ubuntu-software, ahem, gnome-software is using the system libfwupd and not the snapped version?

Reply:
I do like the simplicity of the FWUPD_LIBDIR-only approach -- it's just fixing the one problem

Reply:
The snap store is fully confined and talks to the host daemon.  So expect the client library to move with the confined store itself.
@kenvandine am I right that snap store can just update interface at the same time the library version bumps?

Reply:
>  So expect the client library to move with the confined store itself

That seems sane; I think we don't need the snap-specific path then.

Reply:
I've made it a lot simpler, we can dig this back up if indeed snap needs the fix.

Reply:
> The snap store is fully confined and talks to the host daemon. So expect the client library to move with the confined store itself.
> @kenvandine am I right that snap store can just update interface at the same time the library version bumps?

Yes, we build libfwupd into the snap and talk to the host service.  So no dependency on the lib on the host.  So FWUPD_LIBDIR should be fine.